File: js/src/protocol/auth/encrypted.ts
/**
 * ChaCha20-Poly1305 AEAD encrypted connection layer.
 * Mirrors Python opendesk.protocol.auth.encrypted.
 *
 * Per-direction counters are used as nonces; never transmitted on wire.
 * Uses Node.js built-in crypto (chacha20-poly1305, requires Node 18+).
 */

import crypto from "crypto";
import { Connection, ConnectionClosed } from "../connection.js";

const NONCE_BYTES = 12;
const TAG_BYTES = 16;
const KEY_BYTES = 32;

function makeNonce(counter: number): Buffer {
  const n = Buffer.alloc(NONCE_BYTES);
  // Write counter as 12-byte big-endian
  const hi = Math.floor(counter / 0x1_0000_0000);
  const lo = counter >>> 0;
  n.writeUInt32BE(0, 0);
  n.writeUInt32BE(0, 4);
  n.writeUInt32BE(hi, 4);
  n.writeUInt32BE(lo, 8);
  return n;
}

function chacha20Encrypt(key: Buffer, nonce: Buffer, plaintext: Buffer): Buffer {
  const cipher = crypto.createCipheriv("chacha20-poly1305", key, nonce, {
    authTagLength: TAG_BYTES,
  } as Parameters<typeof crypto.createCipheriv>[3]);
  const enc = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([enc, tag]);
}

function chacha20Decrypt(key: Buffer, nonce: Buffer, ciphertext: Buffer): Buffer {
  const data = ciphertext.subarray(0, -TAG_BYTES);
  const tag = ciphertext.subarray(-TAG_BYTES);
  const decipher = crypto.createDecipheriv("chacha20-poly1305", key, nonce, {
    authTagLength: TAG_BYTES,
  } as Parameters<typeof crypto.createDecipheriv>[3]);
  decipher.setAuthTag(tag);
  try {
    return Buffer.concat([decipher.update(data), decipher.final()]);
  } catch {
    throw new ConnectionClosed("frame integrity check failed (AEAD tag mismatch)");
  }
}

export class EncryptedConnection extends Connection {
  private inner: Connection;
  private sendKey: Buffer;
  private recvKey: Buffer;
  private sendCounter = 0;
  private recvCounter = 0;
  private _closed = false;
  readonly peerPublic?: Buffer;

  constructor(
    inner: Connection,
    opts: { sendKey: Buffer; recvKey: Buffer; peerPublic?: Buffer },
  ) {
    super();
    if (opts.sendKey.length !== KEY_BYTES || opts.recvKey.length !== KEY_BYTES) {
      throw new Error("session keys must be 32 bytes");
    }
    this.inner = inner;
    this.sendKey = opts.sendKey;
    this.recvKey = opts.recvKey;
    this.peerPublic = opts.peerPublic;
  }

  async send(data: Buffer): Promise<void> {
    if (this._closed) throw new ConnectionClosed();
    const ct = chacha20Encrypt(this.sendKey, makeNonce(this.sendCounter), data);
    this.sendCounter++;
    await this.inner.send(ct);
  }

  async recv(): Promise<Buffer> {
    if (this._closed) throw new ConnectionClosed();
    const ct = await this.inner.recv();
    const pt = chacha20Decrypt(this.recvKey, makeNonce(this.recvCounter), ct);
    this.recvCounter++;
    return pt;
  }

  async close(): Promise<void> {
    if (this._closed) return;
    this._closed = true;
    await this.inner.close().catch(() => {});
  }
}

File: python/opendesk/protocol/auth/encrypted.py
"""A :class:`Connection` wrapper that AEAD-encrypts every frame.

After a successful handshake (:func:`pair_*` / :func:`auth_*`), the two peers
agree on a pair of session keys — one for each direction.  Every protocol
frame thereafter goes through :class:`EncryptedConnection`, which:

* on ``send`` — increments a per-direction counter, uses it as the AEAD
  nonce, and encrypts the plaintext with ChaCha20-Poly1305;
* on ``recv`` — does the inverse with its own counter.

The counter is **not** transmitted on the wire — both sides know what number
to use next.  A mismatch (replay, drop, reorder) fails AEAD verification and
the connection is closed.  Because the underlying transport is reliable +
ordered (TCP under WebSocket), this is safe.
"""

from __future__ import annotations

import contextlib
from typing import Optional

try:
    from cryptography.exceptions import InvalidTag
    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
except ImportError as _exc:  # pragma: no cover
    raise ImportError(
        "cryptography is required: pip install 'opendesk[remote]'"
    ) from _exc

from opendesk.protocol.connection import Connection, ConnectionClosed


_NONCE_BYTES = 12  # ChaCha20-Poly1305


def _nonce(counter: int) -> bytes:
    if counter < 0 or counter >= (1 << 96):  # pragma: no cover
        raise OverflowError("AEAD nonce counter exhausted")
    return counter.to_bytes(_NONCE_BYTES, "big")


class EncryptedConnection(Connection):
    """Wraps another :class:`Connection` with AEAD per-frame encryption.

    The wrapped Connection is owned: closing the wrapper closes the inner.
    """

    KEY_BYTES = 32

    def __init__(
        self,
        inner: Connection,
        *,
        send_key: bytes,
        recv_key: bytes,
        peer_public: Optional[bytes] = None,
    ) -> None:
        if len(send_key) != self.KEY_BYTES or len(recv_key) != self.KEY_BYTES:
            raise ValueError("session keys must be 32 bytes")
        self._inner = inner
        self._send = ChaCha20Poly1305(send_key)
        self._recv = ChaCha20Poly1305(recv_key)
        self._send_counter = 0
        self._recv_counter = 0
        self._peer_public = peer_public
        self._closed = False

    @property
    def peer_public(self) -> Optional[bytes]:
        return self._peer_public

    async def send(self, data: bytes) -> None:
        if self._closed:
            raise ConnectionClosed("send on closed connection")
        ct = self._send.encrypt(_nonce(self._send_counter), data, None)
        self._send_counter += 1
        await self._inner.send(ct)

    async def recv(self) -> bytes:
        if self._closed:
            raise ConnectionClosed("recv on closed connection")
        ct = await self._inner.recv()
        try:
            pt = self._recv.decrypt(_nonce(self._recv_counter), ct, None)
        except InvalidTag as exc:
            await self.aclose()
            raise ConnectionClosed("frame integrity check failed") from exc
        self._recv_counter += 1
        return pt

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        with contextlib.suppress(Exception):
            await self._inner.aclose()
